File: 2586-longest-square-streak-in-an-array/longest-square-streak-in-an-array.java
class Solution {
    public int longestSquareStreak(int[] nums) {
        Arrays.sort(nums);  // Sort the array
        int[] arr = new int[100001];  // Track streak length for each number
        int max = 0;

        for (int n : nums) {
            int root = (int) Math.sqrt(n);
            if (root * root != n) {
                arr[n] = 1;  // No predecessor, start new streak
            } else {
                arr[n] = arr[root] + 1;  // Extend streak from root
            }
            max = Math.max(max, arr[n]);  // Track maximum streak length
        }

        return max == 1 ? -1 : max;  // If no streaks longer than 1, return -1
    }
}

File: 2586-longest-square-streak-in-an-array/longest-square-streak-in-an-array.cpp
auto init = []() { 
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(0);
    return 0;
} ();

class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        unordered_map<long, int> streak;
        sort(nums.begin(), nums.end());
        
        int ans = 0;
        
        for(int i = 0; i < nums.size(); ++i) {
            long sq = (long)nums[i] * nums[i];
            if (streak.find(nums[i]) != streak.end()) {
                streak[sq] = streak[nums[i]] + 1;
            } else {
                streak[sq] = 1;
            }
            ans = max(ans, streak[sq]);
        }
        
        return ans == 1 ? -1 : ans;
    }
};
